src/cunumeric/sort/sort.cu

```diff
@@ -0,0 +1,691 @@
+      if (argsort) {
+        merged_indices = create_buffer<int64_t>(merged_size);
+        // merge with key/value
+        auto p_indices1       = source1.indices.ptr(0);
+        auto p_indices2       = source2.indices.ptr(0);
+        auto p_merged_indices = merged_indices.ptr(0);
+        thrust::merge_by_key(thrust::cuda::par.on(stream),
+                             p_values1,
+                             p_values1 + source1.size,
+                             p_values2,
+                             p_values2 + source2.size,
+                             p_indices1,
+                             p_indices2,
+                             p_merged_values,
+                             p_merged_indices);
+        CHECK_CUDA(cudaStreamSynchronize(stream));
+        source1.indices.destroy();
+      } else {
+        thrust::merge(thrust::cuda::par.on(stream),
+                      p_values1,
+                      p_values1 + source1.size,
+                      p_values2,
+                      p_values2 + source2.size,
+                      p_merged_values);
+        CHECK_CUDA(cudaStreamSynchronize(stream));
+      }
+
+      source1.values.destroy();
+      source2.values.destroy();
+      source2.indices.destroy();
+
+      merge_buffers[pos].values  = merged_values;
+      merge_buffers[pos].indices = merged_indices;
+      merge_buffers[pos].size    = merged_size;
+    }
+  }
+  return merge_buffers[0];
+}
+
+template <LegateTypeCode CODE, int32_t DIM>
+struct SortImplBody<VariantKind::GPU, CODE, DIM> {
+  using VAL = legate_type_of<CODE>;
+
+  void operator()(const Array& input_array,
+                  Array& output_array,
+                  const Pitches<DIM - 1>& pitches,
+                  const Rect<DIM>& rect,
+                  const size_t volume,
+                  const bool argsort,
+                  const Legion::DomainPoint global_shape,
+                  const bool is_index_space,
+                  const Legion::DomainPoint index_point,
+                  const Legion::Domain domain,
+                  const std::vector<comm::Communicator>& comms)
+  {
+    AccessorRO<VAL, DIM> input = input_array.read_accessor<VAL, DIM>(rect);
+
+    size_t my_rank = getRank(domain, index_point);
+
+#ifdef DEBUG_CUNUMERIC
+    std::cout << "GPU(" << my_rank << "): local size = " << volume << ", dist. = " << is_index_space
+              << ", index_point = " << index_point << ", domain/volume = " << domain << "/"
+              << domain.get_volume() << ", dense = " << input.accessor.is_dense_row_major(rect)
+              << ", argsort. = " << argsort << std::endl;
+#endif
+
+    assert(rect.empty() || input.accessor.is_dense_row_major(rect));
+
+    auto stream = get_cached_stream();
+
+    const size_t sort_dim_size = DIM == 1 ? volume : global_shape[DIM - 1];
+
+    // make a copy of the input
+    auto dense_input_copy = create_buffer<VAL>(volume, Legion::Memory::Kind::GPU_FB_MEM);
```

Comment:
Like I said for the CPU variant, I think you can save a copy for the single GPU case if you directly copy the input to the output store and sort it in place.